src/util/parallel/data.ml

```diff
@@ -0,0 +1,219 @@
+      }
+
+    let to_list hm =
+      Array.fold_left (fun acc bucket ->
+          match Atomic.get bucket with
+          | None -> acc
+          | Some bucket -> acc @ Bucket.to_list bucket
+        ) [] (Atomic.get hm.buckets)
+
+    let to_seq hm =
+      let bucket_seq = Array.to_seq (Atomic.get hm.buckets) |> Seq.filter_map Atomic.get in
+      Seq.flat_map Bucket.to_seq bucket_seq
+
+    let find_option (hm : t) (key : key): value Option.t =
+      let hash = abs @@ H.hash key in
+      let bucket = Array.get (Atomic.get hm.buckets) (hash mod (Atomic.get hm.nr_buckets)) in
+      Option.bind (Atomic.get bucket) (fun bucket -> Bucket.find_option bucket key)
+
+    let find hm key =
+      match find_option hm key with
+      | None -> raise Not_found
+      | Some value -> value
+
+    let mem hm key =
+      Option.is_some @@ find_option hm key
+
+
+    let rec find_create (hm : t) (key : H.t) =
+      let rec find_create_inner hm key hash buckets =
+        let bucket = Array.get buckets (hash mod (Atomic.get hm.nr_buckets)) in
+        match Atomic.get bucket with
+        | None ->
+          let new_bucket = Bucket.create key in
+          let success = Atomic.compare_and_set bucket None (Some new_bucket) in
+          if success then (
+            (new_bucket.value, true) 
+          )
+          else find_create_inner hm key hash buckets 
+        | Some bucket -> 
+          Bucket.find_create bucket key
+      in
+      let current_generation = Atomic.get hm.resize_generation in
+      let hash = abs @@ H.hash key in
+      let value, was_created = find_create_inner hm key hash (Atomic.get hm.buckets) in
+      if (current_generation mod 2 = 0) && (Atomic.get hm.resize_generation = current_generation || not was_created) then (
+        if (Atomic.get hm.nr_elements >= Atomic.get hm.nr_buckets * 2) then (
+          resize hm;
+        );
+        if (was_created) then Atomic.incr hm.nr_elements;
+        (value, was_created)
+      )
+      else (
+        while (Atomic.get hm.resize_generation = current_generation) do () (* spin *)
+        done;
+        find_create hm key)
+
+
+    and resize hm =
+      let current_generation = Atomic.get hm.resize_generation in
+      if ((current_generation mod 2 = 0) && Atomic.compare_and_set hm.resize_generation current_generation (current_generation+1)) then (
+
+        let old_size = Atomic.get hm.nr_buckets in
+        let new_size = old_size * 2 in
+
+        (* Note that we need a new atomic for each element, so we need Array.init *)
+        let new_buckets = Array.init new_size (fun _ -> Atomic.make None) in
+        let rec rehash_bucket (bucket: Bucket.t option Atomic.t) =
+          match Atomic.get bucket with
+          | None -> ()
+          | Some {key; value; next} -> 
+            begin
+              let hash = abs @@ H.hash key in
+              let new_location = Array.get new_buckets (hash mod new_size) in
+              let _ = match Atomic.get new_location with
```

Comment:
As far as I can see this always returns `()`? So maybe writing it using `;` makes that more clear?